### DISCUSSION TITLE
Version 0.4

### DISCUSSION THREAD

**Discussion by etemesi254** [etemesi254/zune-image#146] [Releases]
Hi, pinging parties I think are relevant(hope I'm not disturbing you).

I'm planning to release version 0.4 of the crates which contains some small niceties but I thought it would be nice to inform you in advance since it contains changes you people requested

@Shnatsel  there are some new decoders as usual I'd like to release

zune-hdr: HDR decoding and encoding, output is f32, as of testing with a simple sample it comes at about 2.7X faster than image-rs/hdr decoder(it doesn't convert HDR to LDR, and has some nice optimizations inside it)

zune-bmp: BMP decoding also landed, support for multiple format types, haven't yet checked speed (but bmp is a most of the times a simple format,checking for speed would be mainly checking for memory copying speeds), but it's quite fast since it has some nice use of iterators and some few tricks that speed up the base case

The two decoders were fuzz tested in CI but additional testing in case there are more available samples wouldn't hurt.

@emilk I'd like you to check if log issues were fixed, there is a new option to disable logs completely for all crates which was added in  https://github.com/etemesi254/zune-image/pull/141 may be of help to you.

Also in case there are benchmarks for image decoding in rerun-io i'd love to see how the library stacks up

@kvark Since you requested for release I'd like to cross check if the library works correctly with you downstream dependencies.

Some things won't be in the release
zune-image: Not yet ready, needs some api decisions
zune-imageprocs: Will be tied to zune-image, also some operations are slightly corrupt and it lacks support for some common image operations, so not ready
zune-python: (Python bindings to the crate, quite fast), not yet as it would expose zune-image.

All changes can be found in  https://github.com/etemesi254/zune-image/blob/dev/CHANGELOG.md

Thank you for filing issues and sending pull requests.

**Comment by kvark** [comment]
Thank you for expedited release schedule! I'm going to be talking about Blade on https://github.com/rust-gamedev/meetup/issues/13, and it would be great to have a new Blade release for this. It depends on Zune libraries, where you landed a few fixes since the last release that I need. Your latest code generally works good for me. Here are the issues:

One problem I see currently is that there is a ton of dependencies dragged by `zune-imageprocs`. I'm only using it for simple things, like Y-flip and resize (for mipmap generation). It happens to depend on `zune-image`, which in turns carries all features/dependencies by default. This is very unfortunate. I wouldn't expect `zune-imageprocs` to depend on any specific image format crates, even transitively.

The other problem, which is not as critical but annoying, is that image compression (e.g. to DXT) is very slow under Debug. I'm seeing 40 seconds per image, which is a sign something is very wrong. Could happen with extensive use of iterators, for example.

**Reply by Shnatsel** [reply]
Such low-level code will always be slow in debug mode. There is nothing that can be reasonably done about it on the source code level.

However, you can override the profile for a specific dependency, so that the crate responsible for DXT always compiles with `-Copt-level=3 ` even in debug mode: https://doc.rust-lang.org/cargo/reference/profiles.html#overrides This will have to be applied in the Blade's `Cargo.toml`.

**Reply by kvark** [reply]
Sorry for bringing this up, texture compression is done with [texpresso](https://github.com/jansol/texpresso) - it has nothing to do with Zune.

**Reply by etemesi254** [reply]
3d04292ae51260d8929fbd276bea525442f63b7e fixes dependency issue in `zune-imageprocs`.

2. We currently don't have resize code, it's a huge undertaking to have a nice resize impl with good speeds, e.g [stb's SIMD's resize](https://github.com/nothings/stb/blob/master/stb_image_resize2.h) is about 10K lines of code.

3. `zune-imageprocs` is one of the heaviest dependencies as it generates optimized code for a lot of operations, takes about 6s to compile on my machine in release, if all you depend on it simple operations, an alternative would be to just have the functions on blade. 
Eg the code for [flip is one of the simplest](https://github.com/etemesi254/zune-image/blob/dev/crates/zune-imageprocs/src/flip.rs) to implement instead of pulling another dependency

**Reply by kvark** [reply]
Got it, thank you for addressing this! Looking forward to see the release on crates :)

**Reply by kvark** [reply]
I see that most of the crates are published, thank you!
Please don't forget about "zune-imageprocs"

**Reply by etemesi254** [reply]
those will take some time as i finish on documentation

**Reply by kvark** [reply]
I hope we can make it before Saturday

**Reply by kvark** [reply]
Is it still blocked on documentation? Technically adding docs is not a breaking change and can be followed up with

**Reply by etemesi254** [reply]
Sorry for the delay, had a really busy week.

Almost done, currently all separate image decoders are released into crates-io https://crates.io/users/etemesi254

Finalizing on imageprocs and image which should be there in about 12 hours

**Reply by etemesi254** [reply]
done

**Reply by etemesi254** [reply]
fixing some small things but 0.4 crates were uploaded

**Reply by etemesi254** [reply]
fixing warnings
